src/regex.cpp

```diff
@@ -191,7 +191,9 @@ BOOST_REGEX_DECL void BOOST_REGEX_CALL put_mem_block(void* p)
 
 #else
 
-#ifdef BOOST_HAS_THREADS
+#if defined(BOOST_REGEX_MEM_BLOCK_CACHE_LOCK_FREE)
+mem_block_cache block_cache;
```

Comment:
We have a problem here: if a regex object is constructed and used at program startup then mem_block_cache may be used uninitialised (since order of initialization is undefined).   For std::atomic we can statically initialize, as with the old code, for boost:atomic we can't unfortunately :(

Only workaround I can think of at present is to replace the global variable with a function returning a reference to the cache:

mem_block_cache& get_cache()
{
   static mem_block_cache cache;
   return cache;
}

Which would also require changing a lot of other code, and making sure the above function is called at program startup to avoid race-conditions when it is eventually called.

There is another issue with boost:atomic, a default initialized atomic as in your code above, actually leaves the value uninitialised, not zero initialized as we require (checked on VC14).  Which complicates things even more!  I'm not sure what to do about this at present, any thoughts?